File: bindings/react-native/ios/RememModule.swift
import ExpoModulesCore
import Foundation

/// React Native binding over rememhq-core, via the Expo Modules API.
///
/// Mirrors the API shape of `Memory` in `bindings/swift` — store, recall,
/// search, update, forget, decay, queryKnowledge, entityContext, and the
/// session lifecycle — but exposed as `AsyncFunction`s returning plain
/// JSON-serializable dictionaries/arrays rather than typed Swift structs,
/// since only primitives, arrays, dictionaries, and `Record` types cross
/// the JS bridge automatically.
///
/// One `EngineHandle` is created per `Remem` instance on the JS side (see
/// `src/index.ts`), keyed by a JS-side numeric handle ID, so a single app
/// can open multiple projects/engines concurrently if it wants to.
///
/// Error surfacing: throwing a `RememError` (a plain `enum: Error`,
/// shared with the SPM Swift binding) from an `AsyncFunction` body
/// rejects the JS promise, per Expo's documented behavior. Expo's own
/// `Exception` type can attach a JS-visible `.code` in addition to a
/// message, but its exact protocol requirements couldn't be verified
/// against source in the environment this was written in — rather than
/// guess at property names and risk a build break, this throws
/// `RememError` directly (which now conforms to `LocalizedError`, so at
/// least `localizedDescription` carries the real message). Revisit with
/// `Exception` once this has been verified against a real Expo Modules
/// Core checkout.
public class RememModule: Module {
    /// Live engine handles, keyed by an opaque ID handed back to JS.
    /// Access is confined to the modules-core dispatch queue (the same
    /// queue every `AsyncFunction` body in this module runs on), so a
    /// plain dictionary is safe without extra locking.
    private var engines: [Int: EngineHandle] = [:]
    private var nextHandleId = 0

    public func definition() -> ModuleDefinition {
        Name("Remem")

        // MARK: - Lifecycle

        AsyncFunction("openEngine") { (project: String, dataDir: String?) -> Int in
            let handle = try EngineHandle(project: project, dataDir: dataDir)
            let id = self.nextHandleId
            self.nextHandleId += 1
            self.engines[id] = handle
            return id
        }

        AsyncFunction("closeEngine") { (engineId: Int) -> Void in
            self.engines.removeValue(forKey: engineId)
        }

        // MARK: - Memory operations

        AsyncFunction("store") {
            (engineId: Int, content: String, tagsJSON: String?, importance: Double) -> [String: Any] in
            let handle = try self.requireEngine(engineId)
            let data = try handle.store(
                content: content, tagsJSON: tagsJSON, importance: Float(importance)
            )
            return try Self.decodeObject(data)
        }

        AsyncFunction("recall") {
            (engineId: Int, query: String, limit: Int, filterTagsJSON: String?) -> [Any] in
            let handle = try self.requireEngine(engineId)
            let data = try handle.recall(query: query, limit: limit, filterTagsJSON: filterTagsJSON)
            return try Self.decodeArray(data)
        }

        AsyncFunction("search") {
            (engineId: Int, query: String, limit: Int, filterTagsJSON: String?) -> [Any] in
            let handle = try self.requireEngine(engineId)
            let data = try handle.search(query: query, limit: limit, filterTagsJSON: filterTagsJSON)
            return try Self.decodeArray(data)
        }

        AsyncFunction("update") {
            (
                engineId: Int, id: String, content: String?, importance: Double,
                tagsJSON: String?
            ) -> [String: Any] in
            let handle = try self.requireEngine(engineId)
            // importance < 0 is the "leave unchanged" sentinel, matching
            // the convention rememhq.h itself uses for remem_update.
            let importanceArg: Float? = importance < 0 ? nil : Float(importance)
            let data = try handle.update(
                id: id, content: content, importance: importanceArg, tagsJSON: tagsJSON
            )
            return try Self.decodeObject(data)
        }

        AsyncFunction("forget") { (engineId: Int, id: String, mode: String) -> Bool in
            let handle = try self.requireEngine(engineId)
            guard let forgetMode = ForgetMode(rawValue: mode) else {
                throw RememError.engine("Unknown forget mode: '\(mode)'")
            }
            return try handle.forget(id: id, mode: forgetMode)
        }

        AsyncFunction("decay") { (engineId: Int, factor: Double) -> Int in
            let handle = try self.requireEngine(engineId)
            return try handle.decay(factor: Float(factor))
        }

        // MARK: - Knowledge graph

        AsyncFunction("queryKnowledge") {
            (engineId: Int, subject: String?, predicate: String?, object: String?) -> [Any] in
            let handle = try self.requireEngine(engineId)
            let data = try handle.queryKnowledge(subject: subject, predicate: predicate, object: object)
            return try Self.decodeArray(data)
        }

        AsyncFunction("entityContext") { (engineId: Int, entity: String) -> [Any] in
            let handle = try self.requireEngine(engineId)
            let data = try handle.entityContext(entity: entity)
            return try Self.decodeArray(data)
        }

        // MARK: - Sessions

        AsyncFunction("startSession") { (engineId: Int, id: String) -> Void in
            let handle = try self.requireEngine(engineId)
            try handle.createSession(id: id)
        }

        AsyncFunction("endSession") { (engineId: Int, id: String) -> Bool in
            let handle = try self.requireEngine(engineId)
            return try handle.endSession(id: id)
        }

        AsyncFunction("getSession") { (engineId: Int, id: String) -> [String: Any]? in
            let handle = try self.requireEngine(engineId)
            guard let data = try handle.getSession(id: id) else { return nil }
            return try Self.decodeObject(data)
        }

        AsyncFunction("listSessions") { (engineId: Int, limit: Int) -> [Any] in
            let handle = try self.requireEngine(engineId)
            let data = try handle.listSessions(limit: limit)
            return try Self.decodeArray(data)
        }

        AsyncFunction("consolidate") {
            (engineId: Int, sessionId: String, model: String?) -> [String: Any] in
            let handle = try self.requireEngine(engineId)
            let data = try handle.consolidate(sessionId: sessionId, model: model)
            return try Self.decodeObject(data)
        }
    }

    // MARK: - Helpers

    private func requireEngine(_ id: Int) throws -> EngineHandle {
        guard let handle = engines[id] else {
            throw RememError.engine(
                "No open engine with handle \(id) (was it already closed?)"
            )
        }
        return handle
    }

    /// Decodes JSON `Data` (always a UTF-8 JSON object or array, per
    /// rememhq.h's documented return conventions) into a JS-bridgeable
    /// dictionary. Deliberately bypasses Memory's typed Codable decoding
    /// — only primitives/arrays/dictionaries cross the Expo bridge
    /// automatically, so going through JSONSerialization instead of
    /// RememCoding.decoder is the right layer here, not a shortcut.
    private static func decodeObject(_ data: Data) throws -> [String: Any] {
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw RememError.decodingFailed(
                "Expected a JSON object, got: \(String(data: data, encoding: .utf8) ?? "<non-utf8>")"
            )
        }
        return object
    }

    private static func decodeArray(_ data: Data) throws -> [Any] {
        guard let array = try JSONSerialization.jsonObject(with: data) as? [Any] else {
            throw RememError.decodingFailed(
                "Expected a JSON array, got: \(String(data: data, encoding: .utf8) ?? "<non-utf8>")"
            )
        }
        return array
    }
}

File: bindings/react-native/android/src/main/java/io/remem/expo/RememModule.kt
package io.remem.expo

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONArray
import org.json.JSONObject

/**
 * Android side of the React Native binding over rememhq-core, via the
 * Expo Modules API. Mirrors `RememModule.swift`'s `AsyncFunction` surface
 * exactly — see that file's doc comment for the full design rationale
 * (the engineId-keying scheme, the -1 importance sentinel, etc.), which
 * applies here unchanged.
 *
 * Unlike iOS, where `EngineHandle`'s C interop required manually
 * decoding raw JSON `Data` via `JSONSerialization`, Kotlin's
 * `org.json.JSONObject`/`JSONArray` are natively supported return types
 * for Expo's Kotlin bridge (confirmed in Expo's own native-module
 * tutorial), so [NativeBridge]'s raw JSON `String` results are simply
 * wrapped in a `JSONObject`/`JSONArray` constructor call here, with no
 * separate decode step needed.
 *
 * Error surfacing: every `NativeBridge` function throws a plain
 * `RuntimeException` (from the JNI layer, via `env.throw_new`) on
 * failure. Per Expo's documented behavior, an `AsyncFunction` body that
 * throws rejects the JS promise automatically — same mechanism, same
 * caveat about `Exception`/`.code` not being used here, as
 * `RememModule.swift`'s own doc comment explains for iOS.
 */
class RememModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Remem")

    // -----------------------------------------------------------------
    // Lifecycle
    // -----------------------------------------------------------------

    AsyncFunction("openEngine") { project: String, dataDir: String? ->
      NativeBridge.openEngine(project, dataDir)
    }

    AsyncFunction("closeEngine") { engineId: Long ->
      NativeBridge.closeEngine(engineId)
    }

    // -----------------------------------------------------------------
    // Memory operations
    // -----------------------------------------------------------------

    AsyncFunction("store") { engineId: Long, content: String, tagsJson: String?, importance: Double ->
      JSONObject(NativeBridge.store(engineId, content, tagsJson, importance.toFloat()))
    }

    AsyncFunction("recall") { engineId: Long, query: String, limit: Int, filterTagsJson: String? ->
      JSONArray(NativeBridge.recall(engineId, query, limit, filterTagsJson))
    }

    AsyncFunction("search") { engineId: Long, query: String, limit: Int, filterTagsJson: String? ->
      JSONArray(NativeBridge.search(engineId, query, limit, filterTagsJson))
    }

    AsyncFunction("update") {
        engineId: Long,
        id: String,
        content: String?,
        importance: Double,
        tagsJson: String?,
      ->
      JSONObject(NativeBridge.update(engineId, id, content, importance.toFloat(), tagsJson))
    }

    AsyncFunction("forget") { engineId: Long, id: String, mode: String ->
      NativeBridge.forget(engineId, id, mode)
    }

    AsyncFunction("decay") { engineId: Long, factor: Double ->
      NativeBridge.decay(engineId, factor.toFloat())
    }

    // -----------------------------------------------------------------
    // Knowledge graph
    // -----------------------------------------------------------------

    AsyncFunction("queryKnowledge") {
        engineId: Long,
        subject: String?,
        predicate: String?,
        objectValue: String?,
      ->
      JSONArray(NativeBridge.queryKnowledge(engineId, subject, predicate, objectValue))
    }

    AsyncFunction("entityContext") { engineId: Long, entity: String ->
      JSONArray(NativeBridge.entityContext(engineId, entity))
    }

    // -----------------------------------------------------------------
    // Sessions
    // -----------------------------------------------------------------

    AsyncFunction("startSession") { engineId: Long, id: String ->
      NativeBridge.startSession(engineId, id)
    }

    AsyncFunction("endSession") { engineId: Long, id: String ->
      NativeBridge.endSession(engineId, id)
    }

    AsyncFunction("getSession") { engineId: Long, id: String ->
      NativeBridge.getSession(engineId, id)?.let { JSONObject(it) }
    }

    AsyncFunction("listSessions") { engineId: Long, limit: Int ->
      JSONArray(NativeBridge.listSessions(engineId, limit))
    }

    AsyncFunction("consolidate") { engineId: Long, sessionId: String, model: String? ->
      JSONObject(NativeBridge.consolidate(engineId, sessionId, model))
    }
  }
}
